shell/platform/fuchsia/dart_runner/dart_component_controller_v2.cc

```diff
@@ -0,0 +1,489 @@
+  // currently, but that may not always be the case. We should avoid making
+  // this assumption and let the CML file specify this data path instead.
+  const std::string component_name = GetComponentNameFromUrl(url_);
+  data_path_ = "pkg/data/" + component_name;
+
+  // TODO(akbiggs): In what situations is the controller not valid? Should
+  // we log an error?
+  if (controller.is_valid()) {
+    binding_.Bind(std::move(controller));
+    binding_.set_error_handler([this](zx_status_t status) { Kill(); });
+  }
+
+  zx_status_t idle_timer_status =
+      zx::timer::create(ZX_TIMER_SLACK_LATE, ZX_CLOCK_MONOTONIC, &idle_timer_);
+  if (idle_timer_status != ZX_OK) {
+    FX_LOGF(INFO, LOG_TAG, "Idle timer creation failed: %s",
+            zx_status_get_string(idle_timer_status));
+  } else {
+    idle_wait_.set_object(idle_timer_.get());
+    idle_wait_.set_trigger(ZX_TIMER_SIGNALED);
+    idle_wait_.Begin(async_get_default_dispatcher());
+  }
+}
+
+DartComponentControllerV2::~DartComponentControllerV2() {
+  if (namespace_) {
+    fdio_ns_destroy(namespace_);
+    namespace_ = nullptr;
+  }
+  close(stdout_fd_);
+  close(stderr_fd_);
+}
+
+bool DartComponentControllerV2::SetUp() {
+  // Name the thread after the url of the component being launched.
+  zx::thread::self()->set_property(ZX_PROP_NAME, label_.c_str(), label_.size());
+  Dart_SetThreadName(label_.c_str());
+
+  if (!CreateAndBindNamespace()) {
+    return false;
+  }
+
+  if (SetUpFromAppSnapshot()) {
+    FX_LOGF(INFO, LOG_TAG, "%s is running from an app snapshot", url_.c_str());
+  } else if (SetUpFromKernel()) {
+    FX_LOGF(INFO, LOG_TAG, "%s is running from kernel", url_.c_str());
+  } else {
+    FX_LOGF(ERROR, LOG_TAG,
+            "Could not find a program in %s. Was data specified"
+            " correctly in the component manifest?",
+            url_.c_str());
+    return false;
+  }
+
+  return true;
+}
+
+bool DartComponentControllerV2::CreateAndBindNamespace() {
+  if (!start_info_.has_ns()) {
+    FX_LOG(ERROR, LOG_TAG, "Component start info does not have a namespace.");
+    return false;
+  }
+
+  const zx_status_t ns_create_status = fdio_ns_create(&namespace_);
+  if (ns_create_status != ZX_OK) {
+    FX_LOGF(ERROR, LOG_TAG, "Failed to create namespace: %s",
+            zx_status_get_string(ns_create_status));
+  }
+
+  dart_utils::RunnerTemp::SetupComponent(namespace_);
+
+  // Bind each directory in start_info's namespace to the controller's namespace
+  // instance.
+  for (auto& ns_entry : *start_info_.mutable_ns()) {
+    // TODO(akbiggs): Under what circumstances does a namespace entry not have a
+    // path or directory? Should we log an error for these?
+    if (!ns_entry.has_path() || !ns_entry.has_directory()) {
+      continue;
+    }
+
+    if (ns_entry.path() == kTmpPath) {
+      // /tmp is covered by the local memfs.
+      continue;
+    }
+
+    // TODO(akbiggs): Why is it okay/desirable to transfer ownership of these?
```

Comment:
The c++ implementation of handles uses RAII to keep the handle open. If we do not take ownership of them they will go out of scope and the handle will be closed meaning that the directory will no longer be valid. We could clone them but since we are not using them elsewhere this is just extra work we are putting on the kernel.

Reply:
Done.